### PR TITLE
[TT-5296] Redis race conditions, refactoring

### PR DESCRIPTION
**Pull request by titpetric**
## Description

- Upgrading our redis library used fixes race conditions.
- Fixing logic errors in ConnectToRedis, adds explicit cancellation
- PubSub issues with over-complicated implementation, cancellation
- DisableRedis and ConnectToRedis logic errors

## Related Issue

This is a root cause of some flaky tests, panics.

## Motivation and Context

Efficiency, correctness, stability.

## How This Has Been Tested

I run all CI tests in a loop 100 times, after disabling flaky tests (separate PR). Some test cases needed to be fixed with the change in implementation here.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [x] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [x] When updating library version must provide reason/explanation for this update.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Check your code additions will not fail linting checks:
  - [x] `go fmt -s`
  - [x] `go vet`

### CODE REVIEW

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4040/merge
Commit:  
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2312135736)

**Comment by Tyk-ITS** [comment]
**API tests result: success** :white_check_mark: 
Branch used: refs/pull/4040/merge
Commit: 1695eceeed515ccf634fabc01fc4750ef70bd6b4 
Triggered by: pull_request (@titpetric)
[Execution page](https://github.com/TykTechnologies/tyk/actions/runs/2337670731)

**Comment by titpetric** [comment]
Merged with https://github.com/TykTechnologies/tyk/pull/4035

File: gateway/auth_manager.go [hunk 1]
```diff
@@ -63,15 +63,15 @@ func (b *DefaultSessionManager) ResetQuota(keyName string, session *user.Session
 	}).Info("Reset quota for key.")
 
 	rateLimiterSentinelKey := RateLimitKeyPrefix + keyName + ".BLOCKED"
+
 	// Clear the rate limiter
-	go b.store.DeleteRawKey(rateLimiterSentinelKey)
+	b.store.DeleteRawKey(rateLimiterSentinelKey)
```
Comment (urbanishimwe) [inline gateway/auth_manager.go:68]: I assume the boolean returned is for if ops was successfully. could we use the return value(like logging). same for line 70, 74.

**Pull request closed without merging**